tests/unit/emulatemb.cxx

```diff
@@ -33,31 +34,295 @@ static const int right          = 0x04;
+  {
+    bool keyEvent;
+    rfb::Point pos;
+    int mask;
+
+    // Middle Button Emulation with Modifier (ALT+Left_click) middle button emulation test code.
+    rdr::U32 keySym;
+    rdr::U32 keyCode;
+    bool down;
+  };
+
+  void handleKeyPress(int keyCode, rdr::U32 keySym);
+  void handleKeyRelease(int keyCode);
+  void filterPointerEvent(const rfb::Point& pos, int buttonMask) {
+    EmulateMB::filterPointerEvent(pos, buttonMask);
+  }
 
   std::vector<PointerEventParams> results;
+  void printResults();
+  int printed;
 };
 
+void TestClass::printResults()
+{
+  for (std::vector<PointerEventParams>::iterator it = results.begin(); it != results.end(); ++it)
+    printf("    sym=%d code=%d down=%s\n", it->keySym, it->keyCode, it->down ? "True" : "False");
+}
+
 void TestClass::sendPointerEvent(const rfb::Point& pos, int buttonMask)
 {
-  PointerEventParams params;
+  PointerEventParams params = {};
+  params.keyEvent = false;
   params.pos = pos;
   params.mask = buttonMask;
+
+  results.push_back(params);
+}
+
+void TestClass::handleKeyPress(int keyCode, rdr::U32 keySym)
+{
+  if (!filterKeyPress(keyCode, keySym))
+    writeKeyEvent(keySym, keyCode, true);
+}
+
+void TestClass::handleKeyRelease(int keyCode)
+{
+  if (!filterKeyRelease(keyCode))
+    writeKeyEvent(0, keyCode, false);
+}
+
+void TestClass::writeKeyEvent(rdr::U32 keySym, rdr::U32 keyCode, bool down)
+{
+  PointerEventParams params = {};
+  params.keyEvent = true;
+  params.keySym = keySym;
+  params.keyCode = keyCode;
+  params.down = down;
+
   results.push_back(params);
 }
 
 #define ASSERT_EQ(expr, val) if ((expr) != (val)) { \
   printf("FAILED on line %d (%s equals %d, expected %d)\n", __LINE__, #expr, (int)(expr), (int)(val)); \
+  test.printResults(); \
   return; \
 }
 
+#define PRESS true
+#define RELEASE false
+
+#define ASSERT_KEY(DOWN, CODE, SYM) do { \
+  ASSERT_EQ(test.results[i].down, DOWN); \
+  ASSERT_EQ(test.results[i].keyCode, CODE); \
+  if ((DOWN) == PRESS) {ASSERT_EQ(test.results[i].keySym, SYM);} \
+  ASSERT_EQ(test.results[i].pos.x, 0); \
+  ASSERT_EQ(test.results[i].pos.y, 0); \
+  ASSERT_EQ(test.results[i].mask, 0); \
+  ++i; \
+} while (0)
+
+#define ASSERT_MOUSE(X, Y, MASK) do { \
+  ASSERT_EQ(test.results[i].pos.x, X); \
+  ASSERT_EQ(test.results[i].pos.y, Y); \
+  ASSERT_EQ(test.results[i].mask, MASK); \
+  ASSERT_EQ(test.results[i].down, 0); \
+  ASSERT_EQ(test.results[i].keyCode, 0); \
+  ASSERT_EQ(test.results[i].keySym, 0); \
+  ++i; \
+} while (0)
+
+// Test press & release of 'a', press & release cmd-a.
+void testMBEMKeyPress(bool enabled)
+{
+  TestClass test;
+
+  printf("%s(%s): ", __func__, enabled ? "Enabled" : "Disabled");
+
+  emulateMiddleButton.setParam(false);
+  emulateMiddleButtonMod.setParam(enabled);
+
+  test.handleKeyPress(a_code, a_sym);
+  test.handleKeyRelease(a_code);
+
+  test.handleKeyPress(cmd_code, cmd_sym);
+  test.handleKeyPress(a_code, a_sym);
+  test.handleKeyRelease(a_code);
+  test.handleKeyRelease(cmd_code);
```

Comment:
Try to split things like these to separate test cases. If things go wrong it makes it easier to see what is broken.